File: packages/lhc/test/fixtures/lifecycle.ts
// The Epic 04 lifecycle script (Story 4, Flow 5): every built v1 surface
// driven once, in PI-extension call order, against one real SDK
// configuration with deterministic inference callbacks — zero network, zero
// model calls, no mocks anywhere. The script owns the sequence and the phase
// names (tech design Flow 5); assertions live in the driving tests, so the
// replay, teardown, and process legs all run THIS sequence rather than a
// re-described one (DD-8).
//
// Phases: create | intake | drain | status | compact1 | pull1 | inspect1 |
// mutate | rebuild | health2 | compact2 | pull2 | materialize.
//
// Mode is background — the production posture the PI extension will run —
// and the drain/rebuild phases settle through `drainSettled`, the real
// background awaitable. The post-mutation health snapshot (AC-5.2's
// "cleared set pending") is deterministic because every SDK operation here
// is synchronous SQLite chained through microtasks, while a background
// drain pass always starts on a macrotask (scheduler runLoop defers through
// setImmediate): the edit → delete → health reads inside the mutate phase
// run as one unbroken microtask cascade and complete before the poked drain
// can claim its first item. The rebuild phase then yields via drainSettled
// and the drain runs to empty.
import { join } from "node:path";
import {
  createDeterministicInferenceCallbacks,
  initLhc,
  type BatchResult,
  type InferenceConfig,
  type CompactReceipt,
  type DerivationReportEntry,
  type HealthReport,
  type InspectOverview,
  type Lhc,
  type MessageEventInput,
  type MutationResult,
  type OpResult,
  type PullResult,
  type ViewContentsReport,
  type ViewStatus,
} from "../../src/index.js";
import { validEvent, type TempStore } from "./index.js";

// ── the one SDK configuration (AC-5.1) ────────────────────────────
//
// Deterministic callbacks injected at construction (the only direct-callback
// path since the CLI retired), background mode, zero-backoff retry so a
// deterministic run never arms a wake timer, the Epic 03 fixture's pinned
// chunk policy (12 fixed-shape turns cut into 4 chunks, c1–c3 closed), and
// a named "lifecycle" profile so compact runs profile-addressed (story
// scope: "compact (profile)") with all three lower bands non-empty on this
// thread (Epic 03 Story 2's reference gradient).
export const LIFECYCLE_PROFILE = {
  name: "lifecycle",
  lowerBound: 400,
  percentages: { full: 25, smooth: 16, detailed: 10, brief: 49 },
} as const;

// Epic 05 TC-4.2: the capstone replays this exact sequence with the real
// inference adapter — the inference-callback slot is the one swap point; the
// sequence, phase names, and every other configuration value stay owned here
// so the deterministic leg and the real leg run THE SAME lifecycle.
export function createLifecycleSdk(inference?: InferenceConfig): Lhc {
  return initLhc({
    ...(inference !== undefined
      ? { inference }
      : { inferenceCallbacks: createDeterministicInferenceCallbacks() }),
    mode: "background",
    retry: { budget: 3, backoffBaseMs: 0, backoffCapMs: 0 },
    chunkPolicy: { targetProjectedTokens: 90, maxProjectedTokens: 4400 },
    view: {
      profiles: [{ ...LIFECYCLE_PROFILE, percentages: { ...LIFECYCLE_PROFILE.percentages } }],
      // Low threshold so the status checkpoint genuinely recommends the
      // compact the sequence performs next.
      compactThreshold: 300,
    },
  });
}

// ── the recorded conversation ─────────────────────────────────────
//
// Same shape and text profile as the Epic 03 derived-thread fixture (12
// turns, tool-heavy middle 5–8, two tool runs each) so the pinned chunk
// policy cuts the same 4-chunk shape — but with explicit per-run
// idempotency keys so a replay on a fresh thread records byte-identical
// event payloads regardless of any process-global fixture counters.
const TURN_COUNT = 12;
const TOOL_HEAVY_TURNS = new Set([5, 6, 7, 8]);
const TURNS_PER_BATCH = 3; // multi-turn batches (story scope)

// Mutation targets: both in the post-compact tail (compact point at t9's
// start under the lifecycle profile), so the second compact's pull shows the
// edit verbatim and the deletion's absence — the tail is full fidelity.
export const EDIT_TARGET = { turnId: "t12", kind: "user_prompt" } as const;
export const EDITED_MESSAGE_TEXT =
  "turn 12 revised: drop area 12 and re-check area 5 instead";
export const DELETE_TARGET = { turnId: "t10", kind: "assistant_text" } as const;
export const DELETED_MESSAGE_TEXT = "findings for area 10";

function turnEvents(turn: number): MessageEventInput[] {
  let seq = 0;
  const key = (): string => {
    seq += 1;
    return `lc-t${turn}-e${seq}`;
  };
  const events: MessageEventInput[] = [
    validEvent("user_prompt", {
      idempotencyKey: key(),
      payload: { text: `turn ${turn}: please investigate area ${turn}` },
    }),
    validEvent("assistant_thinking", {
      idempotencyKey: key(),
      payload: { text: `considering what area ${turn} contains` },
    }),
  ];
  if (TOOL_HEAVY_TURNS.has(turn)) {
    for (const run of [1, 2]) {
      const toolCallId = `call-lc-${turn}-${run}`;
      events.push(
        validEvent("tool_call", {
          idempotencyKey: key(),
          payload: {
            toolCallId,
            toolName: "read_file",
            arguments: { path: `area-${turn}/file-${run}.txt` },
          },
        }),
        validEvent("tool_result", {
          idempotencyKey: key(),
          payload: {
            toolCallId,
            content: `contents of area-${turn}/file-${run}.txt: detail ${turn}.${run} with enough text to summarize`,
            isError: false,
          },
        }),
      );
    }
  }
  events.push(
    validEvent("assistant_text", {
      idempotencyKey: key(),
      payload: { text: `findings for area ${turn}` },
    }),
    validEvent("turn_end", { idempotencyKey: key() }),
  );
  return events;
}

function intakeBatches(): MessageEventInput[][] {
  const batches: MessageEventInput[][] = [];
  for (let first = 1; first <= TURN_COUNT; first += TURNS_PER_BATCH) {
    const batch: MessageEventInput[] = [];
    for (let turn = first; turn < first + TURNS_PER_BATCH && turn <= TURN_COUNT; turn += 1) {
      batch.push(...turnEvents(turn));
    }
    batches.push(batch);
  }
  return batches;
}

// ── run shape ─────────────────────────────────────────────────────

export interface LifecyclePhases {
  create: OpResult<{ threadId: string; filePath: string }>;
  intake: Array<OpResult<BatchResult>>;
  drain: { settled: true };
  status: OpResult<ViewStatus>;
  compact1: OpResult<CompactReceipt>;
  pull1: OpResult<PullResult>;
  inspect1: {
    overview: OpResult<InspectOverview>;
    view: OpResult<ViewContentsReport>;
    health: OpResult<HealthReport>;
  };
  mutate: {
    editedMessageId: string;
    deletedMessageId: string;
    edit: OpResult<MutationResult>;
    delete: OpResult<MutationResult>;
    // Captured inside the mutate phase, before the background drain's first
    // macrotask: the cleared-set-pending evidence (AC-5.2).
    healthAfterMutate: OpResult<HealthReport>;
    messagesNotReady: OpResult<DerivationReportEntry[]>;
    turnsNotReady: OpResult<DerivationReportEntry[]>;
  };
  rebuild: { settled: true };
  health2: OpResult<HealthReport>;
  compact2: OpResult<CompactReceipt>;
  pull2: OpResult<PullResult>;
  materialize: OpResult<{ writtenPath: string }>;
}

export type LifecycleCheckpoint = "inspect1" | "health2" | "materialize";

export interface LifecycleOptions {
  // Distinguishes the thread file and materialize target so replay and
  // teardown legs run side by side in one store (same dir ⇒ same header cwd).
  name?: string;
  // TC-5.3: a fresh initLhc between the intake / compact / mutation /
  // render phase groups — every group boundary is a quiescent point (no
  // queued work, no running drain), so the abandoned instance can never act
  // after its group ends.
  freshSdkBetweenGroups?: boolean;
  // TC-5.4: spawned-CLI parity probes at the three read checkpoints.
  onCheckpoint?: (
    checkpoint: LifecycleCheckpoint,
    ctx: { sdk: Lhc; filePath: string },
  ) => Promise<void>;
  // Epic 05 TC-4.2: run the sequence on the model-call inference path instead
  // of deterministic inference callbacks. Same configuration otherwise (see
  // createLifecycleSdk).
  inference?: InferenceConfig;
}

export interface LifecycleRun {
  filePath: string;
  outPath: string;
  threadId: string;
  phases: LifecyclePhases;
}

function expectOk<T>(result: OpResult<T>, phase: string): T {
  if (!result.ok) {
    throw new Error(`lifecycle ${phase} failed: ${result.error.code} — ${result.error.reason}`);
  }
  return result.value;
}

// ── the sequence ──────────────────────────────────────────────────

export async function runLifecycle(
  store: TempStore,
  opts: LifecycleOptions = {},
): Promise<LifecycleRun> {
  const name = opts.name ?? "lifecycle";
  const filePath = store.threadPath(name);
  const outPath = join(store.dir, `${name}-session.jsonl`);
  const ref = { filePath };

  let sdk = createLifecycleSdk(opts.inference);
  const nextGroup = (): void => {
    // The fresh instance is the same configuration assembled again — the
    // continuity proof is that no phase depends on the prior instance's
    // memory, only on the thread file.
    if (opts.freshSdkBetweenGroups) sdk = createLifecycleSdk(opts.inference);
  };
  const checkpoint = async (cp: LifecycleCheckpoint): Promise<void> => {
    if (opts.onCheckpoint !== undefined) await opts.onCheckpoint(cp, { sdk, filePath });
  };

  // ── group 1: create → intake → drain → status ──
  const create = await sdk.threads.newThread({
    filePath,
    registryPath: store.registryPath,
  });
  const threadId = expectOk(create, "create").threadId;

  const intake: Array<OpResult<BatchResult>> = [];
  for (const batch of intakeBatches()) {
    const sent = await sdk.intakeStream.messageEvents(ref, batch);
    intake.push(sent);
    expectOk(sent, "intake");
  }

  // Background drain settles (story scope): the intake pokes scheduled the
  // work; this is the await that lets the scheduler run it to empty.
  await sdk.drainSettled(ref);
  const drain = { settled: true } as const;

  const status = await sdk.threadView.status(ref);
  expectOk(status, "status");

  // ── group 2: compact1 → pull1 → inspect1 ──
  nextGroup();
  const compact1 = await sdk.threadView.compact(ref, { profile: LIFECYCLE_PROFILE.name });
  expectOk(compact1, "compact1");
  const pull1 = await sdk.threadView.pull(ref);
  expectOk(pull1, "pull1");
  const inspect1 = {
    overview: await sdk.inspect.overview(ref),
    view: await sdk.inspect.view(ref),
    health: await sdk.inspect.health(ref),
  };
  expectOk(inspect1.overview, "inspect1.overview");
  expectOk(inspect1.view, "inspect1.view");
  expectOk(inspect1.health, "inspect1.health");
  await checkpoint("inspect1");

  // ── group 3: mutate → rebuild → health2 ──
  nextGroup();
  const listed = expectOk(await sdk.messages.listMessages(ref), "mutate.list");
  const editTarget = listed.find(
    (record) => record.kind === EDIT_TARGET.kind && record.turnId === EDIT_TARGET.turnId,
  );
  const deleteTarget = listed.find(
    (record) => record.kind === DELETE_TARGET.kind && record.turnId === DELETE_TARGET.turnId,
  );
  if (editTarget === undefined || deleteTarget === undefined) {
    throw new Error("lifecycle invariant: edit/delete targets not found in the record");
  }
  // Edit, delete, and the pending-state reads run in one microtask cascade
  // (see module header): the snapshot is taken before the poked background
  // drain starts its first pass.
  const edit = await sdk.messages.edit(ref, {
    messageId: editTarget.messageId,
    content: EDITED_MESSAGE_TEXT,
  });
  expectOk(edit, "mutate.edit");
  const deleted = await sdk.messages.deleteMessage(ref, { messageId: deleteTarget.messageId });
  expectOk(deleted, "mutate.delete");
  const mutate: LifecyclePhases["mutate"] = {
    editedMessageId: editTarget.messageId,
    deletedMessageId: deleteTarget.messageId,
    edit,
    delete: deleted,
    healthAfterMutate: await sdk.inspect.health(ref),
    messagesNotReady: await sdk.messages.report(ref, { notReady: true }),
    turnsNotReady: await sdk.turns.report(ref, { notReady: true }),
  };
  expectOk(mutate.healthAfterMutate, "mutate.health");
  expectOk(mutate.messagesNotReady, "mutate.messagesNotReady");
  expectOk(mutate.turnsNotReady, "mutate.turnsNotReady");

  // Rebuild drains: the mutations' pokes already scheduled the cascade
  // rebuilds on THIS instance's scheduler; settle is the production await.
  await sdk.drainSettled(ref);
  const rebuild = { settled: true } as const;

  const health2 = await sdk.inspect.health(ref);
  expectOk(health2, "health2");
  await checkpoint("health2");

  // ── group 4: compact2 → pull2 → materialize ──
  nextGroup();
  const compact2 = await sdk.threadView.compact(ref, { profile: LIFECYCLE_PROFILE.name });
  expectOk(compact2, "compact2");
  const pull2 = await sdk.threadView.pull(ref);
  expectOk(pull2, "pull2");
  const materialize = await sdk.threadView.materialize(ref, { path: outPath });
  expectOk(materialize, "materialize");
  await checkpoint("materialize");

  return {
    filePath,
    outPath,
    threadId,
    phases: {
      create,
      intake,
      drain,
      status,
      compact1,
      pull1,
      inspect1,
      mutate,
      rebuild,
      health2,
      compact2,
      pull2,
      materialize,
    },
  };
}

File: packages/lhc-py/tests/fixtures/lifecycle.py
"""Ported from packages/lhc/test/fixtures/lifecycle.ts. Phase 1.

Pure data (profile, mutation targets, conversation builders) are REAL.
SDK-driving helpers (`create_lifecycle_sdk`, `run_lifecycle`) are skeletons.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Literal

from lhc.intake_stream import BatchResult, MessageEventInput
from lhc.messages import EditInput, MessageReportOpts, MutationResult, RemoveInput
from lhc.sdk import Lhc, init_lhc
from lhc.shared_tech.derivation import (
    ChunkPolicyConfig,
    DerivationGuards,
    DerivationReportEntry,
    SdkConfig,
    ToolResultConfig,
)
from lhc.shared_tech.deterministic import create_deterministic_inference_callbacks
from lhc.shared_tech.errors import OpResult
from lhc.shared_tech.inference_types import InferenceConfig
from lhc.shared_tech.inspect import HealthReport, InspectOverview, ViewContentsReport
from lhc.shared_tech.view import (
    CompactReceipt,
    LlmRequestContext,
    PartialViewProfilePercentages,
    SdkViewConfig,
    ViewProfileOverride,
    ViewStatus,
)
from lhc.thread_view import CompactOpts, MaterializeOpts
from lhc.threads import NewThreadResult
from lhc.turns import TurnReportOpts

if TYPE_CHECKING:
    from . import TempStore


# ── the one SDK configuration (AC-5.1) ────────────────────────────


@dataclass(frozen=True, slots=True)
class LifecycleProfilePercentages:
    full: int
    smooth: int
    detailed: int
    brief: int


@dataclass(frozen=True, slots=True)
class LifecycleProfile:
    name: str
    lower_bound: int
    percentages: LifecycleProfilePercentages


LIFECYCLE_PROFILE = LifecycleProfile(
    name="lifecycle",
    lower_bound=400,
    percentages=LifecycleProfilePercentages(full=25, smooth=16, detailed=10, brief=49),
)

_TURN_COUNT = 12
_TOOL_HEAVY_TURNS = frozenset({5, 6, 7, 8})
_TURNS_PER_BATCH = 3


@dataclass(frozen=True, slots=True)
class EditTarget:
    turn_id: str
    kind: str


EDIT_TARGET = EditTarget(turn_id="t12", kind="user_prompt")
EDITED_MESSAGE_TEXT = "turn 12 revised: drop area 12 and re-check area 5 instead"
DELETE_TARGET = EditTarget(turn_id="t10", kind="assistant_text")
DELETED_MESSAGE_TEXT = "findings for area 10"


def _turn_events(turn: int) -> list[MessageEventInput]:
    from . import valid_event

    seq = 0

    def key() -> str:
        nonlocal seq
        seq += 1
        return f"lc-t{turn}-e{seq}"

    events: list[MessageEventInput] = [
        valid_event(
            "user_prompt",
            {"idempotencyKey": key(), "payload": {"text": f"turn {turn}: please investigate area {turn}"}},
        ),
        valid_event(
            "assistant_thinking",
            {"idempotencyKey": key(), "payload": {"text": f"considering what area {turn} contains"}},
        ),
    ]
    if turn in _TOOL_HEAVY_TURNS:
        for run in (1, 2):
            tool_call_id = f"call-lc-{turn}-{run}"
            events.append(
                valid_event(
                    "tool_call",
                    {
                        "idempotencyKey": key(),
                        "payload": {
                            "toolCallId": tool_call_id,
                            "toolName": "read_file",
                            "arguments": {"path": f"area-{turn}/file-{run}.txt"},
                        },
                    },
                )
            )
            events.append(
                valid_event(
                    "tool_result",
                    {
                        "idempotencyKey": key(),
                        "payload": {
                            "toolCallId": tool_call_id,
                            "content": (
                                f"contents of area-{turn}/file-{run}.txt: "
                                f"detail {turn}.{run} with enough text to summarize"
                            ),
                            "isError": False,
                        },
                    },
                )
            )
    events.append(
        valid_event(
            "assistant_text",
            {"idempotencyKey": key(), "payload": {"text": f"findings for area {turn}"}},
        )
    )
    events.append(valid_event("turn_end", {"idempotencyKey": key()}))
    return events


def _intake_batches() -> list[list[MessageEventInput]]:
    batches: list[list[MessageEventInput]] = []
    first = 1
    while first <= _TURN_COUNT:
        batch: list[MessageEventInput] = []
        turn = first
        while turn < first + _TURNS_PER_BATCH and turn <= _TURN_COUNT:
            batch.extend(_turn_events(turn))
            turn += 1
        batches.append(batch)
        first += _TURNS_PER_BATCH
    return batches


@dataclass(frozen=True, slots=True)
class LifecycleDrainPhase:
    settled: Literal[True] = True


@dataclass(frozen=True, slots=True)
class LifecycleInspect1:
    overview: OpResult[InspectOverview]
    view: OpResult[ViewContentsReport]
    health: OpResult[HealthReport]


@dataclass(frozen=True, slots=True)
class LifecycleMutatePhase:
    edited_message_id: str
    deleted_message_id: str
    edit: OpResult[MutationResult]
    delete: OpResult[MutationResult]
    health_after_mutate: OpResult[HealthReport]
    messages_not_ready: OpResult[list[DerivationReportEntry]]
    turns_not_ready: OpResult[list[DerivationReportEntry]]


@dataclass(frozen=True, slots=True)
class LifecycleMaterializeResult:
    written_path: str


@dataclass(frozen=True, slots=True)
class LifecyclePhases:
    create: OpResult[NewThreadResult]
    intake: list[OpResult[BatchResult]]
    drain: LifecycleDrainPhase
    status: OpResult[ViewStatus]
    compact1: OpResult[CompactReceipt]
    llm_context1: OpResult[LlmRequestContext]
    inspect1: LifecycleInspect1
    mutate: LifecycleMutatePhase
    rebuild: LifecycleDrainPhase
    health2: OpResult[HealthReport]
    compact2: OpResult[CompactReceipt]
    llm_context2: OpResult[LlmRequestContext]
    materialize: OpResult[LifecycleMaterializeResult]


LifecycleCheckpoint = Literal["inspect1", "health2", "materialize"]


@dataclass(frozen=True, slots=True)
class LifecycleCheckpointCtx:
    sdk: Lhc
    file_path: str


@dataclass(frozen=True, slots=True)
class LifecycleOptions:
    name: str | None = None
    fresh_sdk_between_groups: bool = False
    on_checkpoint: Callable[[LifecycleCheckpoint, LifecycleCheckpointCtx], Awaitable[None]] | None = None
    inference: InferenceConfig | None = None
    guards: DerivationGuards | None = None


@dataclass(frozen=True, slots=True)
class LifecycleRun:
    file_path: str
    out_path: str
    thread_id: str
    phases: LifecyclePhases


def create_lifecycle_sdk(
    inference: InferenceConfig | None = None,
    guards: DerivationGuards | None = None,
) -> Lhc:
    # Deterministic callbacks injected at construction (the only direct-callback
    # path since the CLI retired), background mode, so a
    # deterministic run never arms a wake timer, the Epic 03 fixture's pinned
    # chunk policy (12 fixed-shape turns cut into 4 chunks, c1–c3 closed), and
    # a named "lifecycle" profile so compact runs profile-addressed (story
    # scope: "compact (profile)") with all three lower bands non-empty on this
    # thread (Epic 03 Story 2's reference gradient).
    # TS: ...(inference !== undefined ? { inference } : { inferenceCallbacks: ... })
    #     ...(guards === undefined ? {} : { guards })
    config = SdkConfig(
        mode="background",
        inference=inference if inference is not None else None,
        inference_callbacks=(
            None if inference is not None else create_deterministic_inference_callbacks()
        ),
        guards=guards if guards is not None else None,
        chunk_policy=ChunkPolicyConfig(target_projected_tokens=90, max_projected_tokens=4400),
        tool_result=ToolResultConfig(
            small_tier_tokens=1, small_target_ratio=0.15, mid_target_ratio=0.04
        ),
        view=SdkViewConfig(
            profiles=[
                ViewProfileOverride(
                    name=LIFECYCLE_PROFILE.name,
                    lower_bound=LIFECYCLE_PROFILE.lower_bound,
                    percentages=PartialViewProfilePercentages(
                        full=LIFECYCLE_PROFILE.percentages.full,
                        smooth=LIFECYCLE_PROFILE.percentages.smooth,
                        detailed=LIFECYCLE_PROFILE.percentages.detailed,
                        brief=LIFECYCLE_PROFILE.percentages.brief,
                    ),
                )
            ],
            # Low threshold so the status checkpoint genuinely recommends the
            # compact the sequence performs next.
            compact_threshold=300,
        ),
    )
    return init_lhc(config)


def _expect_ok(result: OpResult[object], phase: str) -> object:
    if not result.ok:
        raise RuntimeError(
            f"lifecycle {phase} failed: {result.error.code} — {result.error.reason}"
        )
    return result.value


async def run_lifecycle(store: TempStore, opts: LifecycleOptions | None = None) -> LifecycleRun:
    if opts is None:
        opts = LifecycleOptions()
    # TS: opts.name ?? "lifecycle"
    name = "lifecycle" if opts.name is None else opts.name
    file_path = store.thread_path(name)
    out_path = str(Path(store.dir) / f"{name}-session.jsonl")
    ref = {"filePath": file_path}

    sdk = create_lifecycle_sdk(opts.inference, opts.guards)

    def next_group() -> None:
        # The fresh instance is the same configuration assembled again — the
        # continuity proof is that no phase depends on the prior instance's
        # memory, only on the thread file.
        nonlocal sdk
        if opts.fresh_sdk_between_groups:
            sdk = create_lifecycle_sdk(opts.inference, opts.guards)

    async def checkpoint(cp: LifecycleCheckpoint) -> None:
        # TS: if (opts.onCheckpoint !== undefined) await opts.onCheckpoint(...)
        if opts.on_checkpoint is not None:
            await opts.on_checkpoint(cp, LifecycleCheckpointCtx(sdk=sdk, file_path=file_path))

    # ── group 1: create → intake → drain → status ──
    create = await sdk.threads.new_thread(
        {"filePath": file_path, "registryPath": store.registry_path}
    )
    thread_id = _expect_ok(create, "create").thread_id  # type: ignore[union-attr]

    intake: list[OpResult[BatchResult]] = []
    for batch in _intake_batches():
        sent = await sdk.intake_stream.message_events(ref, batch)
        intake.append(sent)
        _expect_ok(sent, "intake")

    # Background drain settles (story scope): the intake pokes scheduled the
    # work; this is the await that lets the scheduler run it to empty.
    await sdk.drain_settled(ref)
    drain = LifecycleDrainPhase(settled=True)

    status = await sdk.thread_view.status(ref)
    _expect_ok(status, "status")

    # ── group 2: compact1 → llmContext1 → inspect1 ──
    next_group()
    compact1 = await sdk.thread_view.compact(ref, CompactOpts(profile=LIFECYCLE_PROFILE.name))
    _expect_ok(compact1, "compact1")
    llm_context1 = await sdk.thread_view.get_llm_request_context(ref)
    _expect_ok(llm_context1, "llmContext1")
    inspect1 = LifecycleInspect1(
        overview=await sdk.inspect.overview(ref),
        view=await sdk.inspect.view(ref),
        health=await sdk.inspect.health(ref),
    )
    _expect_ok(inspect1.overview, "inspect1.overview")
    _expect_ok(inspect1.view, "inspect1.view")
    _expect_ok(inspect1.health, "inspect1.health")
    await checkpoint("inspect1")

    # ── group 3: mutate → rebuild → health2 ──
    next_group()
    listed = _expect_ok(await sdk.messages.list(ref), "mutate.list")  # type: ignore[assignment]
    edit_target = next(
        (
            record
            for record in listed  # type: ignore[union-attr]
            if record.kind == EDIT_TARGET.kind and record.turn_id == EDIT_TARGET.turn_id
        ),
        None,
    )
    delete_target = next(
        (
            record
            for record in listed  # type: ignore[union-attr]
            if record.kind == DELETE_TARGET.kind and record.turn_id == DELETE_TARGET.turn_id
        ),
        None,
    )
    if edit_target is None or delete_target is None:
        raise RuntimeError("lifecycle invariant: edit/delete targets not found in the record")
    # Edit, delete, and the pending-state reads run in one microtask cascade
    # (see module header): the snapshot is taken before the poked background
    # drain starts its first pass.
    edit = await sdk.messages.edit(
        ref,
        EditInput(message_id=edit_target.message_id, content=EDITED_MESSAGE_TEXT),
    )
    _expect_ok(edit, "mutate.edit")
    deleted = await sdk.messages.remove(
        ref, RemoveInput(message_id=delete_target.message_id)
    )
    _expect_ok(deleted, "mutate.delete")
    mutate = LifecycleMutatePhase(
        edited_message_id=edit_target.message_id,
        deleted_message_id=delete_target.message_id,
        edit=edit,
        delete=deleted,
        health_after_mutate=await sdk.inspect.health(ref),
        messages_not_ready=await sdk.messages.report(ref, MessageReportOpts(not_ready=True)),
        turns_not_ready=await sdk.turns.report(ref, TurnReportOpts(not_ready=True)),
    )
    _expect_ok(mutate.health_after_mutate, "mutate.health")
    _expect_ok(mutate.messages_not_ready, "mutate.messagesNotReady")
    _expect_ok(mutate.turns_not_ready, "mutate.turnsNotReady")

    # Rebuild drains: the mutations' pokes already scheduled the cascade
    # rebuilds on THIS instance's scheduler; settle is the production await.
    await sdk.drain_settled(ref)
    rebuild = LifecycleDrainPhase(settled=True)

    health2 = await sdk.inspect.health(ref)
    _expect_ok(health2, "health2")
    await checkpoint("health2")

    # ── group 4: compact2 → llmContext2 → materialize ──
    next_group()
    compact2 = await sdk.thread_view.compact(ref, CompactOpts(profile=LIFECYCLE_PROFILE.name))
    _expect_ok(compact2, "compact2")
    llm_context2 = await sdk.thread_view.get_llm_request_context(ref)
    _expect_ok(llm_context2, "llmContext2")
    materialize = await sdk.thread_view.materialize(ref, MaterializeOpts(path=out_path))
    _expect_ok(materialize, "materialize")
    await checkpoint("materialize")

    return LifecycleRun(
        file_path=file_path,
        out_path=out_path,
        thread_id=thread_id,  # type: ignore[arg-type]
        phases=LifecyclePhases(
            create=create,
            intake=intake,
            drain=drain,
            status=status,
            compact1=compact1,
            llm_context1=llm_context1,
            inspect1=inspect1,
            mutate=mutate,
            rebuild=rebuild,
            health2=health2,
            compact2=compact2,
            llm_context2=llm_context2,
            materialize=materialize,  # type: ignore[arg-type]
        ),
    )
